File: backend/homework/routes/moodle.py
import datetime
import re

import requests
from flask import Blueprint, jsonify, request

from flask_cors import CORS

from homework.database.assignment import Assignment
from homework.database.session import Session
from homework.database.user import User
from homework.database.course import Course
from homework.routes import return_error, to_response
from homework import db, moodle

moodle_bp = Blueprint('moodle', __name__)
CORS(moodle_bp, supports_credentials=True)


@moodle_bp.route('/moodle/get-courses')
def get_courses():
    session_cookie = request.cookies.get('hw_session')
    if not session_cookie:
        return jsonify(return_error('no session')), 401

    session = Session.query.filter_by(id=session_cookie).first()
    if session is None:
        return jsonify(return_error('invalid session')), 401

    user = User.query.filter_by(id=session.user_id).first()
    if user is None:
        return jsonify(return_error('invalid session')), 401

    courses_raw = moodle.get_user_courses(user)
    courses = []
    for course in courses_raw:
        assignments = [assignment.to_dict() for assignment in
                       Assignment.query.filter(Assignment.course == course['id'] and Assignment.from_moodle).all()]

        # convert dates to iso
        for i in range(len(assignments)):
            assignments[i]['dueDate'] = datetime.datetime.strftime(assignments[i]['dueDate'],
                                                                   '%Y-%m-%d')

        print(course)

        course = {
            "id": course['id'],
            'name': course['name'],
            'creator': user.id,
            'assignments': assignments,
            'fromMoodle': True
        }

        courses.append(course)

    return jsonify(to_response(courses))


@moodle_bp.route('/moodle/authenticate', methods=['POST'])
def authenticate():
    session_cookie = request.cookies.get('hw_session')
    if not session_cookie:
        return jsonify(return_error('no session')), 401

    session = Session.query.filter_by(id=session_cookie).first()
    if session is None:
        return jsonify(return_error('invalid session')), 401

    user = User.query.filter_by(id=session.user_id).first()
    if user is None:
        return jsonify(return_error('invalid session')), 401

    base_url, username, password = request.json.get('url'), request.json.get(
        'username'), request.json.get('password')

    if username is None or password is None or base_url is None:
        return jsonify(return_error('missing credentials')), 400

    body = {'username': username, 'password': password}
    auth_request = requests.post(base_url + '/login/token.php?service=moodle_mobile_app', body)
    print(auth_request.json())

    if not auth_request.ok or auth_request.json().get('errorcode') is not None:
        return jsonify(return_error('invalid moodle credentials')), 401

    creds = auth_request.json()

    # get user id
    user_request = requests.post(base_url + '/webservice/rest/server.php?wstoken={}&wsfunction=core_user_get_users_by_field&field=username&values[0]={}&moodlewsrestformat=json'.format(creds['token'], username))

    print(user_request.text)

    user.moodle_url = base_url
    user.moodle_token = creds['token']
    user.moodle_user_id = user_request.json()[0]['id']

    db.session.add(user)
    db.session.commit()

    return jsonify(to_response(user.to_safe_dict()))


@moodle_bp.route('/moodle/get-school-info', methods=['POST'])
def get_school_info():
    school_url = request.json.get('url')
    if school_url is None:
        return jsonify(return_error('no url provided')), 400

    regex = re.compile(
        r'^(?:http|ftp)s?://'  # http:// or https://
        r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # domain...
        r'localhost|'  # localhost...
        r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'  # ...or ip
        r'(?::\d+)?'  # optional port
        r'(?:/?|[/?]\S+)$', re.IGNORECASE)

    if re.match(regex, school_url) is None:
        return jsonify(return_error('invalid url')), 400

    # magic
    url = school_url + "/lib/ajax/service-nologin.php?args=%5B%7B%22index%22%3A0,%22methodname%22%3A%22tool_mobile_get_public_config%22,%22args%22%3A%5B%5D%7D%5D"

    try:
        r = requests.get(url)
    except Exception as e:
        print('[ - ] error retrieving school info: {}'.format(type(e)))
        return jsonify(return_error('invalid url')), 400

    if not r.ok:
        return jsonify(return_error('server error')), 500

    try:
        data = r.json()[0]['data']
    except KeyError:
        return jsonify(return_error('moodle error')), 500

    return jsonify(to_response(data))

File: backend/routes/moodle.go
package routes

import (
	"encoding/json"
	"github.com/3nt3/homework/db"
	"github.com/3nt3/homework/logging"
	"net/http"
	"net/url"
)

func MoodleAuthenticate(w http.ResponseWriter, r *http.Request) {
	HandleCORSPreflight(w, r)

	user, authenticated, err := getUserBySession(r, false)
	if err != nil {
		logging.WarningLogger.Printf("error getting user by session: %v\n", err)
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"error authenticating"},
		}, 500)
		return
	}

	if !authenticated {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"invalid credentials"},
		}, 401)
		return
	}

	// decode moodle credentials
	type moodleLoginData struct {
		Username string `json:"username"`
		Password string `json:"password"`
		URL      string `json:"url"`
	}

	var loginData moodleLoginData
	if err = json.NewDecoder(r.Body).Decode(&loginData); err != nil {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"bad body (not valid json)"},
		}, 400)
		return
	}

	// make request to moodle
	resp, err := http.PostForm(loginData.URL+"/login/token.php?service=moodle_mobile_app", url.Values{
		"username": {loginData.Username},
		"password": {loginData.Password},
	})

	if err != nil {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"error accessing moodle"},
		}, resp.StatusCode)
		logging.WarningLogger.Printf("error: %v", err)
		return
	}

	// decode response
	type moodleTokenResponse struct {
		Token string `json:"token"`
	}
	tokenResp := moodleTokenResponse{}

	err = json.NewDecoder(resp.Body).Decode(&tokenResp)
	if err != nil {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"error accessing moodle"},
		}, 500)
		logging.WarningLogger.Printf("error: %v", err)
		return
	}

	// get user id
	idResp, err := http.PostForm(loginData.URL+"/webservice/rest/server.php", url.Values{
		"wstoken":    {tokenResp.Token},
		"wsfunction": {"core_user_get_users_by_field"},
		"field":      {"username"},
		"values[0]":  {loginData.Username},
		"moodlewsrestformat": {"json"},
	})
	if err != nil {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"error accessing moodle"},
		}, resp.StatusCode)
		logging.WarningLogger.Printf("error: %v", err)
		return
	}

	var responseData []map[string]interface{}
	err = json.NewDecoder(idResp.Body).Decode(&responseData)
	if err != nil {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"error accessing moodle"},
		}, 500)
		logging.WarningLogger.Printf("error: %v", err)
		return
	}

	idInterface, ok := responseData[0]["id"]
	if !ok {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"error accessing moodle"},
		}, 500)
		logging.WarningLogger.Printf("error: no id returned. data: %+v", responseData)
		return
	}

	id := int(idInterface.(float64))

	updatedUser, err := db.UpdateMoodleData(user, loginData.URL, tokenResp.Token, id, false)
	if err != nil {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"internal server error"},
		}, 500)
		logging.InfoLogger.Printf("error updating user: %v\n", err)
		return
	}
	_ = returnApiResponse(w, apiResponse{Content: updatedUser.GetClean()}, 200)
}

func MoodleGetSchoolInfo(w http.ResponseWriter, r *http.Request) {
	HandleCORSPreflight(w, r)
	type requestStruct struct {
		Url string `json:"url"`
	}

	var requestData requestStruct
	if err := json.NewDecoder(r.Body).Decode(&requestData); err != nil {
		logging.WarningLogger.Printf("invalid json: %v\n", err)
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"bad body"},
		}, http.StatusBadRequest)
		return
	}
	if _, err := url.Parse(requestData.Url); err != nil {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"invalid url"},
		}, http.StatusBadRequest)
		return
	}

	requestURL := requestData.Url + "/lib/ajax/service-nologin.php?args=[{\"index\":0,\"methodname\":\"tool_mobile_get_public_config\",\"args\":[]}]"

	resp, err := http.Get(requestURL)
	if err != nil {
		logging.WarningLogger.Printf("error accessing moodle: %v\n", err)
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"invalid url"},
		}, http.StatusBadRequest)
		return
	}

	if resp.StatusCode != 200 {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"invalid url"},
		}, http.StatusBadRequest)
		return
	}

	var moodleSiteData []map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&moodleSiteData); err != nil || len(moodleSiteData) == 0 {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"invalid url", "moodle returned bad data"},
		}, http.StatusBadRequest)
		return
	}

	relevantData, ok := moodleSiteData[0]["data"]
	if !ok {
		_ = returnApiResponse(w, apiResponse{
			Content: nil,
			Errors:  []string{"invalid url", "moodle returned bad data"},
		}, http.StatusBadRequest)
		return
	}

	_ = returnApiResponse(w, apiResponse{Content: relevantData}, 200)
}